### ISSUE TITLE
Add functionality that allows any warnings/errors generated in the R environment to be read in the Python environment

### ISSUE THREAD

**Issue by lloyddewit** [rpy2/rpy2#894]
I am using rpy2 to call an R function. This R function generates warnings. In my Python code, I want to be able to read any warnings generated by the R function. In my specific case, I want to use pytest to check that the R function raises the expected warnings. But there are many other use cases where a developer may want to read/process R environment warnings in Python code.

In the source code, I see that rpy2 writes R warnings to `rpy2.rinterface_lib.callbacks.logger`. I can therefore add a handler to `logger` to forward its warnings to a file/stream etc.

However, if I do this, then I see that 'logger' only seems to generate the R environment warnings when the application exits. It doesn't seem possible to read/process R environment warnings in Python code.

I can demonstrate the issue with the code below:
```
    from rpy2.rinterface_lib.callbacks import logger as rpy2_logger

    log_file_handler = logging.FileHandler("test.log")
    rpy2_logger.addHandler(log_file_handler)

    rpy2_logger.warning("warning1")
    r_warning = rpy2.rinterface.baseenv["warning"]
    r_warning("warning2")
    rpy2_logger.warning("warning3")
```
I expected the contents of `test.log` to be something like:
```
warning1
R[write to console]: 1: warning2 
warning3
```
However, after the final line of code above executes, the contents of `test.log` is:
```
warning1
warning3
```
It is only when the program exits that the R environment warnings are generated and appended to the log file. The contents of `test.log` is then:
```
warning1
warning3
R[write to console]: Warning messages:

R[write to console]: 1: warning2 
```

**In summary, I request to add functionality that allows any warnings/errors generated in the R environment to be read in the Python environment.**

For example, after the final line of code executes, the contents of `test.log` should be similar to:
```
warning1
R[write to console]: 1: warning2
warning3
```

I hope this makes sense. Please let me know if you have any questions.
Thank you!

**Comment by lgautier** [comment]
Yes, I have noticed that warnings only appear when exiting the process. I think that it caused by the following nesting:
```
Python ->   cffi  -> Python
         |       ^
         |       |
         + -> R -+
              |
              v
               callback to Python report the warning        
```
The callback to Python to report the warning is technically while Python/cffi think that we have left Python for C. There is likely a stack of warnings that should be flushed but is not until the process exits.

**Comment by lgautier** [comment]
This is definitely primarily on the R side. There is a queur of warnings that should be flushed when returning from R, and a NULL (a sentinel?) creating trouble:

```python
In [6]: %R flush()
R[write to console]: Error in UseMethod("flush") : 
  no applicable method for 'flush' applied to an object of class "NULL"

R[write to console]: In addition: 
R[write to console]: Warning message:

R[write to console]: In withVisible({ :
R[write to console]:  foo

Error in UseMethod("flush") : 
  no applicable method for 'flush' applied to an object of class "NULL"
```

**Comment by lgautier** [comment]
Setting up R's "warn" option to 1 can lead to a workaround.

This gets warnings appear as they happen:
```python
In [1]: %load_ext rpy2.ipython

In [2]: %R options(warn=1)
Out[2]: o{'warn': [0], }

In [3]: %%R
   ...: warning("foo")
   ...: 
   ...: 
R[write to console]: Warning in withVisible({ :
R[write to console]:  foo
```

A custom callback can change how R warnings (and errors, I don't think there is a way to tell) are handled. Adding a handler to `callbacks.logger` could also achieve that.

```python
In [4]: import rpy2.rinterface_lib.callbacks as callbacks

In [5]: warnerrors = []
   ...: def myconsolewrite_warnerror(s):
   ...:     warnerrors.append(s)
   ...: 

In [6]: callbacks.consolewrite_warnerror = myconsolewrite_warnerror

In [7]: %%R
   ...: warning("foo2")
   ...: 
   ...: 

In [8]: warnerrors
Out[8]: ['Warning in withVisible({ :', ' foo2\n']
```

I am not sure using Python warning is really going to work. The warning is a issued by R calling a Python callback. It could look like this though:

```python
In [9]: import warnings
   ...: def myconsolewrite_warnerror(s):
   ...:     warnings.warn(s)
   ...: 

In [10]: callbacks.consolewrite_warnerror = myconsolewrite_warnerror

In [11]: %%R
    ...: warning("foo3")
    ...: 
    ...: 
<ipython-input-9-2cbee6959a91>:3: UserWarning: Warning in withVisible({ :
  warnings.warn(s)
<ipython-input-9-2cbee6959a91>:3: UserWarning:  foo3

  warnings.warn(s)
```

**Comment by lloyddewit** [comment]
@lgautier Just a quick thank you for looking into this.
I'm currently working on a project that doesn't use rpy2 but we'll need rpy2 again in the coming months so this would be a great enhancement.

**Comment by lgautier** [comment]
Same root as issue #226  I thing.

**Comment by lgautier** [comment]
PR #1023 is introducing a forced flush of R warnings after each "R magic" cell. It provides a template for doing the same for R code evaluation outside of the R magic, and a general API for will likely happen in a future release.

**Comment by lloyddewit** [comment]
@lgautier Thank you, I had a quick look at PR #1023 and it looks promising. I am working again on an rpy2 project and may try this approach. We are also planning to use rpy2 for future products. We will want to check the errors returned by R so if a capability like this was added to the API it would be very valuable.
Thanks again for your responsiveness
